### PR TITLE
Fix FORTE test with mingw

### PR DESCRIPTION
**Pull request by kumajaya**

### CODE REVIEW

**Comment by kumajaya** [comment]
Test result:
```
Running 1333 test cases...
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(65): error: in "PermitTests/setTest": check checkBothOutputEvents() has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed
/home/user/forte/sources/org.eclipse.4diac.forte/tests/stdfblib/CFB_TEST_tester.cpp(69): error: in "PermitTests/setTest": check checkForSingleOutputEventOccurence(0) has failed

```

By using different compilers, we may get different FORTE experiences.

File: tests/core/utils/ifSpecBuilderTest.cpp [hunk 1]
```diff
@@ -22,6 +22,16 @@
 using namespace forte::core::util;
 
 
+template<typename IndexType>
+std::tuple<const CStringDictionary::TStringId*, IndexType>
+CStringIdListSpecBuilder::CStringIdListSpecBuilder::build(CMixedStorage &paStorage) const {
```
Comment (kumajaya) [inline tests/core/utils/ifSpecBuilderTest.cpp]: @MartinMelikMerkumians @azoitl This does not appear to be correct for resolving linking errors.

Comment (MartinMelikMerkumians) [reply tests/core/utils/ifSpecBuilderTest.cpp]: Could you please explain a bit more what the issue is? I don't get what you are meaning

Comment (kumajaya) [reply tests/core/utils/ifSpecBuilderTest.cpp]: > Could you please explain a bit more what the issue is? I don't get what you are meaning

It's a copy from src/core/utils/ifSpecBuilder.cpp, without it I got linking error:
```
/home/user/toolchain/xpack-mingw-w64-gcc-13.2.0-1/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld: CMakeFiles/forte_test.dir/objects.a(ifSpecBuilderTest.cpp.obj):ifSpecBuilderTest.cpp:(.text+0x18e3f): undefined reference to `std::tuple<unsigned long long const*, unsigned long long> forte::core::util::CStringIdListSpecBuilder::build<unsigned long long>(forte::core::util::CMixedStorage&) const'
collect2: error: ld returned 1 exit status
make[2]: *** [tests/CMakeFiles/forte_test.dir/build.make:3467: tests/forte_test.exe] Error 1
make[1]: *** [CMakeFiles/Makefile2:2334: tests/CMakeFiles/forte_test.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

Comment (azoitl) [reply tests/core/utils/ifSpecBuilderTest.cpp]: This seems more to be an issue that ifspcecbuilder.cpp is missing in the test. Copying code can not be the solution for this problem.

Comment (kumajaya) [reply tests/core/utils/ifSpecBuilderTest.cpp]: @MartinMelikMerkumians  I just build FORTE with uclibc compiler with:
```
FORTE_TESTS=ON
FORTE_USE_TEST_CONFIG_IN_FORTE=OFF
FORTE_RTTI_AND_EXCEPTIONS=ON
```
and also experience this linking error problem. No problem if I build with FORTE_TESTS=ON pnly, without:
```
FORTE_USE_TEST_CONFIG_IN_FORTE=OFF
FORTE_RTTI_AND_EXCEPTIONS=ON
```

Comment (MartinMelikMerkumians) [reply tests/core/utils/ifSpecBuilderTest.cpp]: I did a little bit of research, and I don't think it's a good idea to replicate the code in the test, as it needs to be maintained separately.

I see one possible solutions:
Move the template method definition from the cpp to the h file, to make it visible to each compilation unit.

However, there are several template method definitions in ifspecbuilder.cpp but I don't know the reason. Perhaps @azoitl knows why. If its for compilation time, perhaps `extern template` can be used to fix this?

**Review by MartinMelikMerkumians — CHANGES_REQUESTED** [review]

File: tests/core/utils/ifSpecBuilderTest.cpp (hunk 1 above)
Comment (kumajaya) [reply tests/core/utils/ifSpecBuilderTest.cpp]: @MartinMelikMerkumians I found this is the only template method definition in the cpp and move it to the h file fix the linker problem.

**Review by MartinMelikMerkumians — APPROVED** [review]

**Review by azoitl — APPROVED** [review]

**Pull request merged**